servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -151,36 +154,71 @@ toApplication ra request respond = do
 -- The accept header check can be performed as the final
 -- computation in this block. It can cause a 406.
 --
-data Delayed :: * -> * where
-  Delayed :: IO (RouteResult a)
-          -> IO (RouteResult ())
-          -> IO (RouteResult b)
-          -> (a -> b -> RouteResult c)
-          -> Delayed c
+data Delayed c = forall captures auth body. Delayed
+  { capturesD :: IO (RouteResult captures)
+  , methodD   :: IO (RouteResult ())
+  , authD     :: IO (RouteResult auth)
+  , bodyD     :: IO (RouteResult body)
+  , serverD   :: (captures -> auth -> body -> RouteResult c)
+  }
```

Comment:
There is a syntax for GADTs with selectors. (See somewhere here: https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/data-type-extensions.html#gadt-style.) I would prefer that.